Add guide about communicating with a frontend - Closes #917 (#925)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/app-development/frontend.adoc

= Communicating with a frontend
Mona Bärenfänger <mona@lightcurve.io>
:toc:
:sectnums:
:imagesdir: ../../../assets/images
// External URLs
:url_reactjs: https://reactjs.org/
// Project URLS
:url_guides_asset: guides/app-development/asset.adoc
:url_guides_module: guides/app-development/module.adoc
:url_guides_setup: guides/app-development/setup.adoc

This guide explains step-by-step, how to build a simple frontend that communicates with a blockchain application built with the Lisk SDK.

.Prerequisites
[NOTE]
====
To follow this guide, the following criteria is assumed:

* a basic blockchain application is already initiated, as explained in the previous guide xref:{url_guides_setup}[].
====

> To interact with the blockchain application conveniently through a browser, we want to build a simple frontend application.

Reviewer: ```suggestion
To interact with the blockchain application conveniently through a browser, it is necessary to build a simple frontend application.
```

Author: ```suggestion
To interact with the blockchain application conveniently through a browser, it is possible to build a simple frontend application.
```

This frontend can be built with any technology stack you feel most comfortable with.
In this example, we will use {url_reactjs}[React.js].

We will use the `@liskhq/lisk-client` package in the frontend application to communicate with the blockchain application.

== Create a new React app

> First let's take a look at the current file structure of the project:

Reviewer: ```suggestion
The current file structure of the project can be seen below:
```

----
├── blockchain-app
│   ├── index.js
│   └── package.json
----

Create a new folder for the frontend on the root level of your project:

[source,bash]
----
npx create-react-app frontend-app
----

This will automatically set up a React project for you with default configurations in a newly created `frontend-app` folder.

----
├── blockchain-app
│   ├── index.js
│   └── package.json
├── frontend-app
│   ├── public/
│   ├── src/
│   ├── README.md
│   └── package.json
----

It is already possible to start the frontend at this point:

[source,bash]
----
cd frontend-app
npm start
----

== Install dependencies

[source,bash]
----
npm i react-router-dom
npm i @liskhq/lisk-client
----

> To use `BigInt` in the frontend, it might be needed to add the following options to the `package.json` file:

Reviewer: ```suggestion
To use `BigInt` in the frontend, it may be required to add the following options to the `package.json` file:
```

.frontend-app/package.json
[source,js]
----
{
  // [...]
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ],
    "env": {
      "es2020": true,
      "browser": true,
      "node": true,
      "mocha": true
    }
  },
  // [...]
}
----

== Create basic components

> Now, let's customize this simple app by creating different components for the first basic functions of the frontend:

Reviewer: ```suggestion
This simple app can be customized by creating different components for the first basic functions of the frontend as shown below:
```

* New account: Generates new account credentials.
* Faucet: A component that sends funds to a specified account from the genesis account.

> * Send transfer transaction: A component that allows to send token from one account to another.

Reviewer: ```suggestion
* Send transfer transaction: A component that allows sending tokens from one account to another.
```

* Account details: Returns details of a user account by address.

=== New account

> A page for generating new accounts allows to conveniently create credentials that can be used in the application.

Reviewer: ```suggestion
A page for generating new accounts that conveniently allows the creation of credentials that can be used in the application.
```

Import `passphrase` and `cryptography` from the `lisk-client` package to create new account credentials.

[source,jsx]
----
import { passphrase, cryptography } from '@liskhq/lisk-client';

const newCredentials = () => {
    const pass = passphrase.Mnemonic.generateMnemonic();
    const keys = cryptography.getPrivateAndPublicKeyFromPassphrase(pass);
    const credentials = {
        address: cryptography.getBase32AddressFromPassphrase(pass),
        binaryAddress: cryptography.getAddressFromPassphrase(pass).toString("hex"),
        passphrase: pass,
        publicKey: keys.publicKey.toString("hex"),
        privateKey: keys.privateKey.toString("hex")
    };
    return credentials;
};

const NewAccount = () => {
    const credentials = newCredentials();
    return (
        <div>
            <h2>Create new account</h2>
            <p>Refresh page to get new credentials.</p>
            <pre>{JSON.stringify(credentials, null, 2)}</pre>
        </div>
    );
}
export default NewAccount;
----

=== Faucet

> The faucet is a component that allows accounts to receive tokens from the genesis account, which holds majority of initial tokens at the start of the Devnet.

Reviewer: ```suggestion
The faucet is a component that allows accounts to receive tokens from the genesis account, which holds the majority of initial tokens at the start of the Devnet.
```

> The faucet is a component that allows accounts to receive tokens from the genesis account, which is the account that hold the vast majority of all tokens at the start of the Devnet.

Reviewer 2: ```suggestion
The faucet is a component that allows accounts to receive tokens from the genesis account, which holds majority of initial tokens at the start of the Devnet.
```

> In a new file `api.js`, the `apiClient` from package `lisk-client` provides an interface for the faucet and other React components to connect to the blockchain application via websocket on port 8888.

Reviewer: ```suggestion
In a new file `api.js`, the `apiClient` from package `lisk-client` provides an interface for the faucet and other React components to connect to the blockchain application via a websocket on port 8888.
```

                </label>
                <label>
                    Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={handleChange} value={state.amount} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            {state.transaction && <7>
                <div>
                    <pre>Transaction: {JSON.stringify(state.transaction, null, 2)}</pre>
                    <pre>Response: {JSON.stringify(state.response, null, 2)}</pre>
                </div>
            }
        </div>
    );
};

export default Faucet;
----

<1> Inside `Faucet.js`, we import the previously defined API client from `api.js`.
<2> `transactions` and `cryptography` from the `lisk-client` package are used to convert the data of the transaction into the correct format.
<3> The passphrase for the genesis account of the Devnet.
<4> The API client is used to create the transaction object based on the inputs in the form below.
<5> After creation, the transaction is submitted to the blockchain application.

> <6> After submitting the transaction and getting back the response, the state of the Faucet component is updated with the transaction object and the API response.

Reviewer: ```suggestion
<6> After submitting the transaction and receiving the response, the state of the Faucet component is updated with the transaction object and the API response.
```

<7> If the transaction object is present in the state, it is displayed below the form on the same page.

=== Send transaction

> Now that we are able to create a new account and receive some initial tokens, let's build a new component that allows to send tokens from an account to another.

Reviewer: ```suggestion
Now that we are able to create a new account and receive some initial tokens, it is possible to build a new component that allows sending tokens from one account to another.
```

> Now that we are able to create a new account and receive some initial tokens for it, let's build a new component that allows to send tokens from an account to another.

Reviewer: ```suggestion
Now that we are able to create a new account and receive some initial tokens, it is possible to build a new component that allows sending tokens from an account to another.
```

Author: ```suggestion
Now that we are able to create a new account and receive some initial tokens, let's build a new component that allows to send tokens from an account to another.
```

To do this, create a new file `Transfer.js`.

> The contents of `Transfer.js` are similar to `Faucet.js`, as we send a transfer transaction on both pages.

Reviewer: ```suggestion
The contents of `Transfer.js` are similar to `Faucet.js`, as a transfer transaction will be sent on both pages.
```

                </label>
                <label>
                    Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={handleChange} value={state.amount} />
                </label>
                <label>
                    Fee:
                        <input type="text" id="fee" name="fee" onChange={handleChange}  value={state.fee} />
                </label>
                <label>
                    Passphrase:
                        <input type="text" id="passphrase" name="passphrase" onChange={handleChange}  value={state.passphrase} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            {state.transaction &&
                <div>
                    <pre>Transaction: {JSON.stringify(state.transaction, null, 2)}</pre>
                    <pre>Response: {JSON.stringify(state.response, null, 2)}</pre>
                </div>
            }
        </div>
    );
}
export default Transfer;
----

<1> Here the transaction gets signed with the passphrase provided in the form.

=== Account details

> As a last component, let's add a page that displays account details by address.

Reviewer: ```suggestion
For the final component, we can add a page that displays the account details by address.
```

        address: '',
        account: {},
    });

    const handleChange = (event) => {
        const { name, value } = event.target;
        updateState({
            ...state,
            [name]: value,
        });
    };

    const handleSubmit = async (event) => {
        event.preventDefault();
        const client = await api.getClient();
        const account = await client.account.get(cryptography.getAddressFromBase32Address(state.address)); <1>
        updateState({
            ...state,
            account: client.account.toJSON(account),
        });
    };

    return (
        <div>
            <h2>Account</h2>
            <p>Get account details by address.</p>
            <form onSubmit={handleSubmit}>
                <label>
                    Address:
                        <input type="text" id="address" name="address" onChange={handleChange} value={state.address} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            <div>
                <pre>Account: {JSON.stringify(state.account, null, 2)}</pre> <2>
            </div>
        </div>
    );
}
export default Account;
----

> <1> Retrieves account details from the blockchain application, based on the address provided.

Reviewer: ```suggestion
<1> Retrieves the account details from the blockchain application, based on the address provided.
```

<2> If `state.account` is not empty, it will display the account details on the same page.

=== Index and navigation

> Now that all the basic components for the frontend are created, let's add a small component for the landing page.

Reviewer: ```suggestion
Now that all the basic components for the frontend are created, a small component for the landing page can be added.
```

            <h3> Explore </h3>
            <li><Link to="/account">Account</Link></li>
          </ul>
        </Route>

        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route path="/new-account">
            <NewAccount />
          </Route>
          <Route path="/faucet">
            <Faucet />
          </Route>
          <Route path="/send-transfer">
            <Transfer />
          </Route>
          <Route path="/account">
            <Account />
          </Route>
        </Switch>
      </div>
    </Router>
  );
}

export default app;
----

In the already existing `index.js` file, the `App.js` component is finally included in the `root` element, which is defined in `index.html`.

.index.js
[source,jsx]
----
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root')
);
----

== View in browser

After completing all the steps above, start the app again:

[source,bash]
----
npm start
----

This should open the app in browser under the URL http://localhost:3000 .

> It is now possible to use the app in browser to create new accounts, fund accounts, see account details of a specific account, and send tokens from one account to another.

Reviewer: ```suggestion
It is now possible to use the app in a browser to create new accounts, fund accounts, view the account details of a specific account, and send tokens from one account to another.
```

=== New account

A page for generating new accounts allows to conveniently create credentials that can be used in the frontend.

Import `passphrase` and `cryptography` from the `lisk-client` package to create new account credentials.

[source,jsx]
----
import { passphrase, cryptography } from '@liskhq/lisk-client';

const newCredentials = () => {
    const pass = passphrase.Mnemonic.generateMnemonic();
    const keys = cryptography.getPrivateAndPublicKeyFromPassphrase(pass);
    const credentials = {
        address: cryptography.getBase32AddressFromPassphrase(pass),
        binaryAddress: cryptography.getAddressFromPassphrase(pass).toString("hex"),
        passphrase: pass,
        publicKey: keys.publicKey.toString("hex"),
        privateKey: keys.privateKey.toString("hex")
    };
    return credentials;
};

const NewAccount = () => {
    const credentials = newCredentials();
    return (
        <div>
            <h2>Create new account</h2>
            <p>Refresh page to get new credentials.</p>
            <pre>{JSON.stringify(credentials, null, 2)}</pre>
        </div>
    );
}
export default NewAccount;
----

=== Faucet

> The faucet is a component that allows accounts to receive tokens from the genesis account, which is the account that hold the vast majority of all tokens at the start of the Devnet.

Reviewer 2: `which holds the majority of initial tokens`

> To send the tokens to the desired account, we use the `apiClient` from the `lisk-client` package and create a new file `api.js` which will provide the api client to the faucet and other React components.

Reviewer 2: L147 & L149 can probably be combined as one.

"In a new file `api.js` the `apiClient` from package `lisk-client`  provides an interface for faucet & other React components to connect to the blockchain application via websocket on port 8888"

Inside of `api.js`, the function `getClient()` is defined which establbishes a WebSocket connection to the blockchain application.

.api.js
[source,js]
----
const { apiClient } = require('@liskhq/lisk-client');
const RPC_ENDPOINT = 'ws://localhost:8888/ws';

let clientCache;

export const getClient = async () => {
    if (!clientCache) {
        clientCache = await apiClient.createWSClient(RPC_ENDPOINT);
    }
    return clientCache;
};
----

> Next, create a new file `Faucet.js`, which stores the React component for the faucet.

Reviewer 2: "which `will` store the React component `of` the faucet"

        updateState({ <6>
            transaction: client.transaction.toJSON(tx),
            address: '',
            amount: '',
            response:response
        });
    }

    return (
        <div>
            <h2>Faucet</h2>
            <p>The faucet transfers tokens from the genesis account to another.</p>
            <form onSubmit={handleSubmit}>
                <label>
                    Address:
                        <input type="text" id="address" name="address" onChange={handleChange} value={state.address} />
                </label>
                <label>
                    Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={handleChange} value={state.amount} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            {state.transaction && <7>
                <div>
                    <pre>Transaction: {JSON.stringify(state.transaction, null, 2)}</pre>
                    <pre>Response: {JSON.stringify(state.response, null, 2)}</pre>
                </div>
            }
        </div>
    );
};

export default Faucet;
----

> <1> Inside of `Faucet.js`, we import the previously defined API client from `api.js`.

Reviewer 2: "Inside `Faucet.js`"

> <2> The `transactions` and `cryptopgraphy` packages from the `lisk-client` package are used to convert the data of the transaction into the correct format.

Reviewer 2: I think we should be consistent if referring `transactions` & `cryptography` as packages (Ref L111)

<3> The passphrase for the genesis account of the Devnet.

> <4> The API client is used to create the transaction object based on the inputs of the form below.

Reviewer 2: "`in` the form below."

> <5> After creation, the transaction is posted to the blockchain application.

Reviewer 2: "the transaction is `submitted`"

> <6> After posting the transaction, the state of the Faucet component is updated with the transaction object and the API response.

Reviewer 2: "After `submitting` the transaction and `getting back the response`, the state of..."

<7> If the transaction objbect is present in the state, it is displayed below the form on the same page.

=== Send transaction

> Now that we are able to create a new account and receive some initial tokens for it, let's built a new component that allows to send tokens from an account to another.

Reviewer 2: `build`

To do this, create a new file `Transfer.js`.

> The contents of `Transfer.js` are similar so `Faucet.js`, as we send a transfer transaction on both pages.

Reviewer 2: "are similar `to`"

> The only difference is, that the sender is not specified to be the genesis account, and instead it can be sent by any account to any account in the network.

Reviewer 2: The only difference is, that the sender is not essentialy a genesis account, but can be any account in the network.

                    Recipient:
                        <input type="text" id="address" name="address" onChange={handleChange} value={state.address} />
                </label>
                <label>
                    Amount (1 = 10^8 tokens):
                        <input type="text" id="amount" name="amount" onChange={handleChange} value={state.amount} />
                </label>
                <label>
                    Fee:
                        <input type="text" id="fee" name="fee" onChange={handleChange}  value={state.fee} />
                </label>
                <label>
                    Passphrase:
                        <input type="text" id="passphrase" name="passphrase" onChange={handleChange}  value={state.passphrase} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            {state.transaction &&
                <div>
                    <pre>Transaction: {JSON.stringify(state.transaction, null, 2)}</pre>
                    <pre>Response: {JSON.stringify(state.response, null, 2)}</pre>
                </div>
            }
        </div>
    );
}
export default Transfer;
----

> <1> Here the transaction gets signed with the passphrase that is provided in the form.

Reviewer 2: "with the passphrase provided in the form"

        address: '',
        account: {},
    });

    const handleChange = (event) => {
        const { name, value } = event.target;
        updateState({
            ...state,
            [name]: value,
        });
    };

    const handleSubmit = async (event) => {
        event.preventDefault();
        const client = await api.getClient();
        const account = await client.account.get(cryptography.getAddressFromBase32Address(state.address)); <1>
        updateState({
            ...state,
            account: client.account.toJSON(account),
        });
    };

    return (
        <div>
            <h2>Account</h2>
            <p>Get account details by address.</p>
            <form onSubmit={handleSubmit}>
                <label>
                    Address:
                        <input type="text" id="address" name="address" onChange={handleChange} value={state.address} />
                </label>
                <input type="submit" value="Submit" />
            </form>
            <div>
                <pre>Account: {JSON.stringify(state.account, null, 2)}</pre> <2>
            </div>
        </div>
    );
}
export default Account;
----

> <1> Receives account details from the blockchain application, based on the provided address.

Reviewer 2: "`Retrieves` account details from the blockchain application, based on the address `provided`."

<2> If `state.account` is not empty, it will display the account details on the same page.

=== Index and navigation

> Now that all the basic components for the frontend are created, let's add a small component which will be the landing page.

Reviewer 2: "let's add a small component `for` the landing page."

} from "react-router-dom";
import Home from './Home';
import NewAccount from './NewAccount';
import Faucet from './Faucet';
import Account from './Account';
import Transfer from './Transfer';

export const app = () => {
  return (
    <Router>
      <div>
        <Route>
          <ul>
            <li><Link to="/">Home</Link></li>
            <hr />
            <h3> Interact </h3>
            <li><Link to="/new-account">New Account</Link></li>
            <li><Link to="/faucet">Faucet</Link></li>
            <li><Link to="/send-transfer">Send Transfer</Link></li>
            <hr />
            <h3> Explore </h3>
            <li><Link to="/account">Account</Link></li>
          </ul>
        </Route>

        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route path="/new-account">
            <NewAccount />
          </Route>
          <Route path="/faucet">
            <Faucet />
          </Route>
          <Route path="/send-transfer">
            <Transfer />
          </Route>
          <Route path="/account">
            <Account />
          </Route>
        </Switch>
      </div>
    </Router>
  );
}

export default app;
----

> In the already existing `index.js` file, the `App.js` component is finally included in the `root` element, which is defined at `index.html`.

Reviewer 2: "defined `in`"

.index.js
[source,jsx]
----
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root')
);
----

> == View in the browser

Reviewer 2: View in browser

> After completing all the above steps, start the app again:

Reviewer 2: "all the steps above"

[source,bash]
----
npm start
----

> This should open the app in the browser under `localhost:3000`.

Reviewer 2: "This should open the app in browser under the URL `localhost:3000`."

> It is now possible to use the browser to create new accounts, fund accounts, see account details of a specific account, and sending tokens from one account to another.

Reviewer 2: "It is now possible to use the `app in` browser to create new accounts, fund accounts, see account details of a specific account, and `send` tokens from one account to another.
